### DISCUSSION TITLE
Overlapping multi component SMARTS matching

### DISCUSSION THREAD

**Discussion by jpfeuffer** [rdkit/rdkit#7216] [Q&A] [answered]
Hi!

I am not sure this is a bug, but Daylight specifies the "dot" as notation for "not *necessarily* connected" which in my opinion implies that the different components as specified in a SMARTS *may* be connected but don't have to.

In RDKit the following yields no matches:

```python
test = "CCO"
testMol = Chem.MolFromSmiles(test)
testSMARTS = "[#8]-*.[#6]-*"
testSMARTSMol = Chem.MolFromSmarts(testSMARTS)
matches = testMol.GetSubstructMatches(testSMARTSMol)
print(matches)
```

while with `test="CCCO"` it does.

**Comment by jasondbiggs** [comment, marked as answer]
The issue is that `testSMARTSMol` has four atoms while `testMol` only has three.  Either of these SMARTS should show the behavior of disconnected pattern matching:

```
"[#8]-*.[#6]"
"[#8]-*.[$([#6]-*)]"
```

**Reply by jpfeuffer** [reply]
Thank you. Hmm I kind of see what you mean, but who "dictates" that those two wildcard atoms are not allowed to be the same and therefore four atoms are necessary in the molecule to be matched?

In any case, since I did not know about your workaround with the recursive SMARTS, I want to see if there is a workaround for my actually a bit bigger problem as well.
Maybe you could help me with your expertise. I would like to have a single reaction SMARTS/SMIRKS, that matches both of my test SMILES in the following snippet.

```python
testSMARTS = "([F:1][#6:2]-[$([#1,*]):3].[Cl][#6:5]-[$([#1,*]):6])>>([F:1][#6:2]-[$([#1,*]):3].[Br][#6:5]-[$([#1,*]):6])"
testSMARTSRxn = AllChem.ReactionFromSmarts(testSMARTS)


test = "CC(C)(F)Cl"
testMol = Chem.MolFromSmiles(test)

ps = testSMARTSRxn.RunReactants((testMol,))
print("Products: ")
print(len(ps)) # 0 = no match

test2 = "CC(F)C(C)Cl"
test2Mol = Chem.MolFromSmiles(test2)

ps2 = testSMARTSRxn.RunReactants((test2Mol,))
print("Products: ")
print(len(ps2)) # 1 product
```

**Reply by greglandrum** [reply]
> Thank you. Hmm I kind of see what you mean, but who "dictates" that those two wildcard atoms are not allowed to be the same and therefore four atoms are necessary in the molecule to be matched?

The RDKit’s rules for SMARTS matching are quite simple: every atom in the SMARTS has to match an atom in the molecule. Recursive SMARTS match a single atom (the atoms “inside” the recursion don’t count there)

**Reply by jpfeuffer** [reply]
Recursive SMARTS seem to have done the trick! Thank you both. I will try this on more complex reactions now.

https://gist.github.com/jpfeuffer/42174d7f8d14b026d9bbd56a0136ffc1#file-untitled1-ipynb
